### ISSUE TITLE
Using a formatter or memory profile stops cache/parallelization from being used

### ISSUE THREAD

**Issue by broksonic21** [rubocop/rubocop#11664]
## Expected behavior

When you run 

`bundle exec rubocop --fail-level W --display-only-fail-level-offenses --cache true -d `

the cache is used per the logs:

```
Use parallel by default.
Running parallel inspection
Loading cache from .rubocop-cache/rubocop_cache/REDACTED/REDACTED/REDACTED_FILE
Loading cache from .rubocop-cache/rubocop_cache/REDACTED/REDACTED/REDACTED_FILE
```

I expected running this

`bundle exec rubocop --fail-level W --display-only-fail-level-offenses --cache true -d --format json`
or
`bundle exec rubocop --fail-level W --display-only-fail-level-offenses --cache true -d --format junit`
or
`bundle exec rubocop --fail-level W --display-only-fail-level-offenses --cache true -d --memory --profile`

would do the same

## Actual behavior

It doesn't however, you get no lines regarding cache or parallel, only scanning

```
Scanning /Users/REDACTED/REDACTED/app/Gemfile
Scanning /Users/REDACTED/REDACTED/app/Rakefile
Scanning /Users/REDACTED/REDACTED/REDACTED_PATH/REDACTED_FILE.rb
Scanning /Users/REDACTED/REDACTED/REDACTED_PATH/REDACTED_FILE.rb
```

## Steps to reproduce the problem

See commands above

## RuboCop version

```
1.47.0 (using Parser 3.2.1.0, rubocop-ast 1.27.0, running on ruby 2.7.7) [arm64-darwin21]
```

**Comment by broksonic21** [comment]
@fatkodima this is the additional findings I came across via https://github.com/rubocop/rubocop/issues/11657#issuecomment-1455152503 , just broken into its own ticket re: cache/parallel impact.

**Comment by fatkodima** [comment]
Regarding `bundle exec rubocop --fail-level W --display-only-fail-level-offenses --cache true -d --memory --profile` - this is expected and the cache is explicitly disabled, because otherwise you would always get incorrect results when profiling and the cache is enabled (it would be very fast and won't actually profile anything).

Regarding the `format` - need to take a look.

**Comment by broksonic21** [comment]
Got it re: memory. Lower-pri than formatters, but might be good to document that in either memory or cache docs.
Neither https://docs.rubocop.org/rubocop/usage/caching.html nor https://docs.rubocop.org/rubocop/usage/profiling.html mention that case.

**Comment by jvlara** [comment]
Hi @koic @fatkodima, is there any update on the issue where using the format option doesn't use the cache?
We have a large Ruby on Rails codebase, and when we run RuboCop with the clang formatter, it takes about 7 minutes.
```sh
bin/rubocop \
  --debug \
  --parallel \
  --extra-details \
  --display-style-guide \
  --fail-level convention \
  --format clang
```
However, when we use a different formatter, the execution time increases to 70 minutes. We suspect that the cache is not being used in this case.

**Comment by jvlara** [comment]
I finally tracked the bug to `$LOADED_FEATURES`—its contents differ between the start and end of the run, so the cache never hits (`lib/rubocop/result_cache.rb#rubocop_extra_features`).

This happens when formatters are lazy-loaded (e.g., via `autoload`).

The quick fix was to eagerly load them in the RuboCop binstub, like this:

```ruby
require "pathname"

ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../../Gemfile",
  Pathname.new(__FILE__).realpath)

require "rubygems"
require "bundler/setup"

# These two made $LOADED_FEATURES differ, this fixes it
require_relative "../gems/rubocop-buk/lib/rubocop-buk.rb"
require_relative "../gems/rubocop-buk/lib/rubocop/buk/jsonl_formatter.rb"

load Gem.bin_path("rubocop", "rubocop")
```

**Comment by Earlopain** [comment]
@jvlara There has been a fix very recently (https://github.com/rubocop/rubocop/pull/14339). Maybe what you encounter is separate  problem but I can't really tell without knowing what version you are using. Since you already tracked down a cause I guess you are on latest, can you confirm?

**Comment by jvlara** [comment]
@Earlopain I think this is a different issue. My logs show that parallel execution is happening, but the cache created by Parallel isn’t used by the main RuboCop process — likely due to differences in $LOADED_FEATURES, as I mentioned earlier.

I'm on RuboCop 1.74.0.

<img width="188" height="54" alt="Image" src="https://github.com/user-attachments/assets/c00112e8-4578-48f6-988f-dfbab67b0bc5" />

I pinpoint the issue by looking at the logs, this were the logs that i get without loading explicitily the formatter 

<img width="1033" height="602" alt="Image" src="https://github.com/user-attachments/assets/f4588451-1647-40da-b62b-34bd04b85934" />

And this were the logs after the binstub hack 

<img width="1040" height="604" alt="Image" src="https://github.com/user-attachments/assets/93dc6176-9ba5-46ca-a10d-64b798477e18" />

Also note the time, the last one takes 7 minutes while the other one i have to cancel it because it was in the 30 minutes mark

**Comment by jvlara** [comment]
I'd try making a PR, but $LOADED_FEATURES is tricky — I'd need to autoload everything that might be loaded before RuboCop starts, or consider changing that behavior entirely.

Then again, the issue might just be my formatter being lazy-loaded and not initialized until the first run. Forcing all formatters to safe_constantize up front might solve it.

Also, adding logs here would help — I had to dig through the code to find the root cause too much, if it using Parallel and then not being able to found the file should give a warning or something .

EDIT: i was overengineering things, a memoize should solve the issue...
